Add support for BERDY_FLOATING_BASE in BerdySparseMAPSolver

traversaro opened the pull request:
Also added a consistency test between `BerdySparseMAPSolver` and `ExtWrenchesAndJointTorquesEstimator` .

traversaro commented:
The only test failing is the octave one. Let me know when you think that this is ready to be merged, so I can regenerate and commit the bindings.

francesco-romano commented on src/core/include/iDynTree/Core/TestUtils.h:
```diff
@@ -311,6 +311,46 @@ namespace iDynTree
+        if( !checkCorrect )
+        {
+            std::cerr << file << ":" << line << " : assertVectorAreEqualWithRelativeTol failure: " << std::endl;
+            printVector("vec1",vec1);
```
Comment for a later PR: I think we could add a newline between the various prints, so we understand when a vector ends and the other starts

traversaro replied:
Yes, I am also always bothered by that.

traversaro commented:
This branch is finally fully working. I rebased away the bindings to have a branch easy to review. @francesco-romano @diegoferigo if you want you can check the latest 5/6 commits, but it would be great to merge this as soon as possible to avoid the branch balkanization.

traversaro commented:
(Bindings need to be regenerated before merging).

francesco-romano commented on src/estimation/src/BerdyHelper.cpp:
```diff
@@ -542,6 +543,11 @@ IndexRange BerdyHelper::getRangeSensorVariable(const SensorType type, const unsi
         sensorOffset += 3*m_sensors.getNrOfSensors(THREE_AXIS_ANGULAR_ACCELEROMETER);
     }
 
+    if( type > THREE_AXIS_FORCE_TORQUE_CONTACT )
```
Sorry.. I probably saw this method a lot of time, but only now I am wondering.. why `type > XXX` ?

traversaro replied:
Basically is quite a crude way to express the logic "if the sensor comes after all the sensors of type XXX in serialization, then you have to use the proper offset when computing the sensor range in the sensor serialization.

francesco-romano commented on src/estimation/src/BerdyHelper.cpp:
```diff
@@ -1270,9 +1300,11 @@ bool BerdyHelper::computeBerdySensorMatrices(SparseMatrix<iDynTree::ColumnMajor>
-                matrixYElements.addDiagonalMatrix(sensorRange,
-                                                  netExtWrenchRange,
-                                                  1);
+                Transform measurementFrame_X_link = (m_link_H_externalWrenchMeasurementFrame[idx].inverse());
```
parenthesis should not be needed (also line 1284)

francesco-romano commented on src/sensors/src/ThreeAxisForceTorqueContactSensor.cpp:
```diff
@@ -0,0 +1,232 @@
+        // We assume that the total forcetorque is given by the sum of pure forces on the z axis measured by the load cells
+        Wrench loadCellWrench = Wrench(LinearForceVector3(0.0, 0.0, loadCellMeasurements(i)*1.0), AngularForceVector3(0.0, 0.0, 0.0));
+        Transform sensor_H_loadCell(Rotation::Identity(), pimpl->m_loadCellLocations[i]);
+        totalWrench = totalWrench + sensor_H_loadCell*loadCellWrench;
```
nitpick: I usually find more compact (but also readable) the `+=` operator

traversaro replied:
There is no `Wrench::operator+=` at the moment. : )

francesco-romano replied:
I did't realise it was on the Wrench object. Ok :)

francesco-romano commented on src/sensors/src/ThreeAxisForceTorqueContactSensor.cpp:
```diff
@@ -0,0 +1,232 @@
+        return zeroVec;
+    }
+
+    // We assume that the location of the load cells is on the XY plane. In this case, the cop is just the average of the locations weighted on the measurements
```
Just a question: this should also match to the result computed by `computeThreeAxisForceTorqueFromLoadCellMeasurements` and properly manipulated, right? (like torque / f_z)
Can it be used as a consistency test?

traversaro replied:
Yes, indeed.

traversaro commented on src/estimation/include/iDynTree/Estimation/BerdyHelper.h:
```diff
@@ -699,6 +723,22 @@ class BerdyHelper
+     */
+    ///@{
+    bool setNetExternalWrenchMeasurementFrame(const LinkIndex lnkIndex, const Transform& link_H_contact);
+    bool getNetExternalWrenchMeasurementFrame(const LinkIndex lnkIndex, Transform& link_H_contact);
```
Good point.

traversaro commented:
@francesco-romano The latest two commits should have addressed your comments. Let me know if they are ok so I can do the final rebase+bindings generation.

francesco-romano commented:
👍  Go with the merge
